pageserve/test_routes.py

```diff
@@ -0,0 +1,68 @@
+"""Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import unittest
+from unittest.mock import patch
+from flask import Flask
+from flask_testing import TestCase
+import app
+
+EXAMPLE_USER = "app_user"
+EXAMPLE_POSTS = ['example', 'posts', 'list']
+EXAMPLE_EVENTS = ['example', 'events', 'list']
+
+
+class TestRoutes(TestCase):
+    def create_app(self):
+        """Creates and returns a Flask instance.
+
+        Required by flask_testing to test templates."""
+        app = Flask(__name__)
+        app.config['TESTING'] = True
+        return app
+
+    def setUp(self):
+        """Set up test client."""
+        self.client = app.app.test_client()
+
+    @patch('app.get_user', return_value=EXAMPLE_USER)
+    @patch('app.has_edit_access', return_value=True)
+    @patch('app.get_posts', return_value=EXAMPLE_POSTS)
+    def test_index(self, mock_posts, mock_access, mock_user):
```

Comment:
Not really sure if this is the best way to patch all these functions for testing, would like some input.

Reply:
I don't like those unused parameters, and pylint should get angry at you too if you run it. Instead, does replacing those with single underscores work? Alternatively, in your body you could also nest patches as context managers:
```python
with patch('app.get_user', return_value=EXAMPLE_USER):
    with patch('app.has_edit_access', return_value=True):
        with patch('app.get_posts', return_value=EXAMPLE_POSTS):
            ...
```
but that just looks ugly.

Reply:
If the concern is pylint, then replacing them with each with single underscores is worse since they're duplicates. Prefixing them with an underscore gets pylint to ignore them, but that doesn't really address the issue of the unused parameters.

Reply:
What if you specify in the patches to replace it with a magic mock with each specific return value. That might allow you to remove the parameters from the function since you are explicitly created the replacement mock objects in place.
```suggestion
    from unittest.mock import MagicMock
    
    @patch('app.get_user', MagicMock(return_value=EXAMPLE_USER))
    @patch('app.has_edit_access', MagicMock(return_value=True))
    @patch('app.get_posts', MagicMock(return_value=EXAMPLE_POSTS))
    def test_index(self):
```

Reply:
That lets me get rid of the parameters, thanks